sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/stringExpressions.scala

```diff
@@ -431,56 +432,233 @@ case class FindInSet(left: Expression, right: Expression) extends BinaryExpressi
 
-  def convert(v: UTF8String): UTF8String = v.trimLeft()
+  override def inputTypes: Seq[AbstractDataType] = Seq.fill(children.size)(StringType)
+  override def dataType: DataType = StringType
+
+  override def nullable: Boolean = children.exists(_.nullable)
+  override def foldable: Boolean = children.forall(_.foldable)
 
   override def prettyName: String = "ltrim"
 
-  override def doGenCode(ctx: CodegenContext, ev: ExprCode): ExprCode = {
-    defineCodeGen(ctx, ev, c => s"($c).trimLeft()")
+  override def eval(input: InternalRow): Any = {
+    val inputs = children.map(_.eval(input).asInstanceOf[UTF8String])
+    if (inputs(0) != null) {
+      if (children.size == 1) {
+        return inputs(0).trimLeft()
+      } else if (inputs(1) != null) {
+        if (inputs(0).numChars > 1) {
+          throw new AnalysisException(s"Trim character '${inputs(0)}' can not be greater than" +
+            s" 1 character.")
+        } else {
+          return inputs(1).trimLeft(inputs(0))
+        }
+      }
+    }
+    null
+  }
+
+    override def doGenCode(ctx: CodegenContext, ev: ExprCode): ExprCode = {
+    if (children.size == 2 &&
+       (! children(0).isInstanceOf[Literal] || children(0).toString.length > 1)) {
+      throw new AnalysisException(s"The trimming parameter should be Literal " +
+        s"and only one character.") }
+
+    val evals = children.map(_.genCode(ctx))
+    val inputs = evals.map { eval =>
+      s"${eval.isNull} ? null : ${eval.value}"
+    }
+    val getTrimLeftFunction = if (children.size == 1) {
+      s"""UTF8String ${ev.value} = ${inputs(0)}.trimLeft();"""
+    } else {
+      s"""UTF8String ${ev.value} = ${inputs(1)}.trimLeft(${inputs(0)});"""
+    }
+
+    ev.copy(evals.map(_.code).mkString("\n") +
+      s"""
+    boolean ${ev.isNull} = false;
+    ${getTrimLeftFunction};
+    if (${ev.value} == null) {
+      ${ev.isNull} = true;
+    }
+    """)
+  }
+
+  override def sql: String = {
+    if (children.size == 1) {
+      val childrenSQL = children.map(_.sql).mkString(", ")
+      s"$prettyName($childrenSQL)"
+    } else {
+      val trimSQL = children(0).map(_.sql).mkString(", ")
+      val tarSQL = children(1).map(_.sql).mkString(", ")
+      s"$prettyName($trimSQL, $tarSQL)"
+    }
   }
 }
 
 /**
- * A function that trim the spaces from right end for given string.
+ * A function that trim the spaces or a character from right end for given string.
  */
 @ExpressionDescription(
   usage = "_FUNC_(str) - Removes the trailing space characters from str.",
-  extended = "> SELECT _FUNC_('    SparkSQL   ');\n '    SparkSQL'")
-case class StringTrimRight(child: Expression)
-  extends UnaryExpression with String2StringExpression {
+  extended = "> SELECT _FUNC_('    SparkSQL   ');\n '    SparkSQL'\n" +
+             "> SELECT _FUNC_('S', 'SSparkSQLS');\n 'SSparkSQL'\n" +
+             "> SELECT _FUNC_(TRAILING 'S' FROM 'SSparkSQLS');\n 'SSparkSQL'")
```

Comment:
Is the TRAILING character here should be constant value or can be another expression?